### ISSUE TITLE
Status code to prevent G Sheet to load csv data

### ISSUE THREAD

**Issue by codingjoe** [metabase/metabase#11974]
**Describe the bug**

If I am to include data into Google Sheets from metabase via the [card-export API](https://github.com/metabase/metabase/blob/master/docs/api-documentation.md#post-apicardcard-idqueryexport-format) via the [IMPORTDATA](https://support.google.com/docs/answer/3093335) function, I get a `Could not fetch URL: ...` error. However, on my server I see the request and the expected 202 response.

It works fine on 0.34.1.

**Logs**

The proxy logs a 202 Accepted on the requested URL. I can not share the exact log, this it reveals a public URL.

**To Reproduce**
1. Go to any card.
2. Share it publicly.
3. Go to Google Sheets
4. Paste the `IMPORTDATA("YOUR_PUBLIC_CSV_URL")` into a cell.

**Expected behavior**

It should simply work, as described in #3772 

**Screenshots**

**Information about your Metabase Installation:**
```json
{
  "browser-info": {
    "language": "de",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "1.8.0_242-heroku-b08",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_242-heroku",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "25.242-b08",
    "os.name": "Linux",
    "os.version": "4.4.0-1062-aws",
    "user.language": "en",
    "user.timezone": "Etc/UTC"
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "googleanalytics"
    ],
    "hosting-env": "heroku",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.6 (Ubuntu 11.6-1.pgdg16.04+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "tag": "v0.34.2",
      "date": "2020-02-04",
      "branch": "release-0.34.x",
      "hash": "dab738c"
    },
    "settings": {
      "report-timezone": "Europe/Berlin"
    }
  }
}
```

**Severity**

 It isn't severe, but annoying for me. I am happy to help however I can.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by codingjoe** [comment]
A quick header diff reveals, that the response code changed:
```diff
< HTTP/1.1 202 Accepted
---
> HTTP/1.1 200 OK
```

**Comment by codingjoe** [comment]
Related discussion in the form: https://discourse.metabase.com/t/importing-csv-data-into-google-sheets-stopped-working/9119

**Comment by flamber** [comment]
Caused by PR #11817
I'm not sure if this is something that Google should fix or we should use a different status code.
A workaround would probably be to use a reverse-proxy and modify the response code.

**Comment by codingjoe** [comment]
> I'm not sure if this is something that Google should fix or we should use a different status code.

Does it really matter? Even if we urge Google to fix it, until then it might be better to go back to a 200 all the poor users.

> A workaround would probably be to use a reverse-proxy and modify the response code.

That's what I did, but that isn't something everyone will be comfortable setting up. It's also a pain if you already have plenty documents that use the `IMPORTDATA` function. Updating them would be a pain.

At the end, I agree that 202 might be the better response since everything is asynchronous. Yet, why even bother starting a response if the request has not been processed.  It's somewhat unfamiliar.

**Comment by SetSails** [comment]
guys! Are you really contemplating whether it might be better to switch back so that Google Sheets integration works again ? Let me make it very simple: HELL YES !!!

This is --- if not THE most important feature of metabase.

I have just upgraded to 0.35.3 just to realize that all our integrations stopped working.

Can you please feed back on what your plan on this is - it's been over 2 months now that you have last written back and forth on this. 

PLEASE get this fixed so that we can go back to the recent version of Metabase.

Thank you so much !!!

**Comment by flamber** [comment]
@SetSails Currently it seems like only 4 people are experiencing this, so that's why there hasn't been much activity. Like with any other issue, use the :+1: on the first post to indicate reaction to an issue.
There has been much work in the new connection handling, but it's not fully completed yet, so upcoming changes might fix issues like this.

**Comment by SetSails** [comment]
Hi @flamber , thank you for getting back so quickly.

Well, are you saying only 4 people are using GSheet integration? Or are others who are using GSheet are not experiencing these issues ?

I really hope this can be addressed and fixed as it really renders metabase completely useless for us unforunately.

Thank you & Best Regards
Fabian

**Comment by codingjoe** [comment]
Hi @SetSails,

The issue is universal. So everyone that uses Google Sheets is experiencing it, yet not everyone is running a version of Metabase that includes this bug. I, too, downgraded to an older version without the bug.

> There has been much work in the new connection handling, but it's not fully completed yet, so upcoming changes might fix issues like this.

@flamber I don't know if it's wise to do regression handling as part of a larger refactoring. Addressing the bug separately might give us a test that could prevent future regression.

Best
-Joe

**Comment by lchachurski** [comment]
Hey @flamber,

Please consider the fact that some "thumbs up" are far greater than 1. We had to downgrade Metabase for a whole organization of nearly 100 people. Fell free to acknowledge the current count of 103. 

Regards

**Comment by nberezowsky** [comment]
I found workaround:

1) You can add this open-source script to your google sheet https://github.com/bradjasper/ImportJSON
2) Use importJSON() function:
![image](https://user-images.githubusercontent.com/9456996/82200990-e604bb00-9929-11ea-821f-7a2b776343ba.png)

**Comment by SetSails** [comment]
Thank you @nberezowsky !! I'll consider giving this a try.

@flamber as you can see the issue is pressing enough that people start coming up with workarounds, even to the point where custom coding is introduced as opposed to leveraging standard features.

Could you please give this topic a reconsideration ? Having to add the importJSON function for every single GSheet where we pull data is a heck of a lot of work.

You'd really make your community happy by fixing this at the root.

Thank you & Best Regards
Fabian

**Comment by vkovalskiy** [comment]
Same here, updated the latest version just to find out all of our CSV imports for Google crashed. We are running on Heroku and I have no idea how to downgrade especially as the DB is already upgraded for the recent Metabase. ImportJSON is a workaround that kills all of the formating so we have to redo everything in sheet configs. This is a fix for only simple CSVs. I am really surprised to know that MB guys did not fix this since Feb. Com'on how come you have simply killed the feature making the behaviour backwards-incompatible within the minor update?

**Comment by lnorbert** [comment]
We're also affected. We have metabase behind HaProxy. Our workaround was to rewrite the response header. We added these two lines to the backend section of metabase:

```
    acl status_202 status 202
    http-response set-status 200 if status_202
```

**Comment by codingjoe** [comment]
@flamber how many more thumbs up or workaround proposals do we need to switch back to a 200 response, that really didn't do any harm? I don't see a point in dragging this out for too long. Would you consider accepting a patch that reverts to the old 200 response code?

**Comment by flamber** [comment]
@codingjoe This is not a simple one-line fix from what I can see. It was part of one of the biggest changes in Metabase (streaming responses). You're more than welcome to submit a PR.

**Comment by SetSails** [comment]
Hi @flamber , what was the intention behind changing this ? what are you trying to accomplish with streaming responses ?

thanks & regards
Fabian

**Comment by flamber** [comment]
@SetSails This #11832 and many other PR descriptions.

**Comment by SetSails** [comment]
Gotcha, thanks @flamber . Maybe a stupid question... but could there not just be a parameter to be appended to a query URL which results in a code 200 instead of 220 to be returned?

Or - maybe even better - query the user agent and return 200 if it spots Google sheets as querying? (At least until GSheet supports the 220's?)

BR Fabian

**Comment by codingjoe** [comment]
@flamber I did end up submitting a patch, reverting everything back to normal 200 responses. However, some parts of the test suite just fail during setup (and seem unrelated to my change). Anyhow, maybe you could already share your opinion until those issues are sorted.

**Comment by flamber** [comment]
@codingjoe :+1: You can ignore the two failing driver tests. Not sure why cypress-2 failed, but it took a very long time to run, so maybe something timed-out on the CI server.
I don't know the consequences, so I'll leave that up to Cam who implemented the streaming logic.

**Comment by shimoju** [comment]
We are also affected. We added this workaround to rewrite the status code by [lua-nginx-module](https://github.com/openresty/lua-nginx-module).

```
header_filter_by_lua '
  if ngx.status == ngx.HTTP_ACCEPTED then
    ngx.status = ngx.HTTP_OK
  end
';
```

**Comment by manhtai** [comment]
> We are also affected. We added this workaround to rewrite the status code by [lua-nginx-module](https://github.com/openresty/lua-nginx-module).
> 
> ```
> header_filter_by_lua '
>   if ngx.status == ngx.HTTP_ACCEPTED then
>     ngx.status = ngx.HTTP_OK
>   end

We also got this issue and solved using this trick, thanks.

For anyone need to know more about `lua-nginx-module`, you can install `nginx-extras` package on Ubuntu/Debian instead of `nginx` and it will be included, no need to compile from source.

**Comment by adamstelle** [comment]
My company's also affected - upgraded versions overnight to find that the 20+ GSheet integrations we rely on are broken. Echo the previous comments that this is a very damaging bug.

**Comment by vkbrihma** [comment]
Also affected!

**Comment by codingjoe** [comment]
@flamber I rebased my patch. CI is passing now. Maybe we should also consider bumping this to P1 considering the 44 up votes and desperate comments.

**Comment by camsaul** [comment]
Bumping this to P1 since there are 50 upvotes now. @robdaemon is going to rebase #12577 and we'll merge the fix in and ship it as part of 0.36.8

**Comment by camsaul** [comment]
Fixed by #13500
